### PR TITLE
Feature order state watcher

### PR DESCRIPTION
**Pull request by perfectmak**
##### Description
This PR implements the Order State Watcher functionality. 
These include the following classes:
- `OrderStateWatcher` - Actually class for adding orders and making a subscription for changes. These class uses the `EventWatcher` and `ExpirationWatcher` to listen for changes and uses the `OrderStateUtils` to compute the relevant state (whether invalid or valid) of all affected orders.

- `OrderStateUtils` - Utility class to compute the relevant state of an Order. It uses all the lazy stores to prevent recomputing of unchanges order params.
- `BalanceAndAllowanceLazyStore` 
- `OrderCollateralPoolAndTokenLazyStore`
- `AbiDecoder` - Utility class to decode event logs gotten from the `EventWatcher`
- `EventWatcher` - Watches for event logs and allows subscribing to these events.

It also contains some fixes for:
- `RemainingFillableCalc` - To calculate check MKT tokens as fees and not collateral token
- `OrderFilledCancelledLazyStore` - To prevent NPE's when deleting none-cached contracts info

##### Checklist
- [x] Linter status: 100% pass
- [x] Changes don't break existing behavior
- [ ] Tests coverage hasn't decreased

##### Refers/Fixes
Fixes: #36

### CODE REVIEW

**Comment by perfectmak** [comment]
@eswarasai I believe this guy is ready for review now. I've also updated the PR's description to provide more details.
Also @cryptomental, please also help take a look, you might have some comments about the changes I made to the `RemainingFillableCalc`.

cc @pelsasser

**Comment by eswarasai** [comment]
@perfectmak -- I'd probably need another day to completely review this. Sorry about the delay. Although I'm curious about the plan to tackle the coverage decrease. I see that it's not part of the bounty but I'd wanted to understand if this is something you can take care of as well? @rojosnow can definitely tip you for any additional work that you've put into. Thanks :)

**Comment by rojosnow** [comment]
@perfectmak If you have the time, please address the coverage issue as well. We can handle the extra time in a tip.

**Comment by perfectmak** [comment]
Yes, @eswarasai I can tackle the decreased coverage. I just wanted to submit a PR first and get feedback on the implementation in time because the task has been taking a while now to complete.
cc @rojosnow

**Comment by perfectmak** [comment]
@eswarasai @rojosnow, I've added more tests to cover for untested cases. The last case I can't really test is the periodic interval calls to `_cleanupAsync()` in the OrderStateWatcherClass, because it is a little complicated to test. But overall I think it has pretty good coverage of most of the scenarios.

Kindly review.

**Comment by rojosnow** [comment]
@perfectmak Thanks! I'll let @eswarasai chime in here.

**Comment by rojosnow** [comment]
@eswarasai What is the status on this PR?

**Comment by eswarasai** [comment]
@rojosnow -- Still under review. Should be able to close it today EOD.

**Comment by rojosnow** [comment]
@eswarasai Cool. Good to close out today?

**Comment by perfectmak** [comment]
It is okay.

**Comment by eswarasai** [comment]
@pelsasser -- This is looking good to me. Would you like to take a look at it before merging since there are whole bunch of changes? Thanks!

File: src/contract_wrappers/ContractWrapper.ts [hunk 1]
```diff
@@ -760,6 +760,15 @@ export class ContractWrapper {
     return tokenContract.approveTx(spenderAddress, amountInBaseUnits).send(txParams);
   }
 
+  /**
+   *
+   * @param marketContractAddress Market Contract Address
+   */
+  public async getCollateralTokenAddressAsync(marketContractAddress: string): Promise<string> {
```
Comment (pelsasser) [inline src/contract_wrappers/ContractWrapper.ts]: @perfectmak - would it be preferable to use the following here?

```
  public async getCollateralTokenAddressAsync(marketContractAddress: string): Promise<string> {
    const contractSet: ContractSet = await this._getContractSetByMarketContractAddressAsync(marketContractAddress);
    return contractSet.collateralToken.address;
  }
```

File: src/order_watcher/EventWatcher.ts [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+  // region Properties
+  // *****************************************************************
+  // ****                     Properties                          ****
+  // *****************************************************************
+  // endregion //Properties
+  // region Public Methods
+  // *****************************************************************
+  // ****                     Public Methods                      ****
+  // *****************************************************************
+
```
Comment (pelsasser) [inline src/order_watcher/EventWatcher.ts]: not needed to merge this, but would be nice to add some documentation to these functions.  We can always do at a later date, but ideally ally public functions at a minimum have some documentation

File: src/order_watcher/RemainingFillableCalc.ts [hunk 3]
```diff
@@ -125,7 +131,7 @@ export class RemainingFillableCalculator {
   private async _getAvailableFeeFunds(accountAddress: string): Promise<BigNumber> {
     const allowance = new BigNumber(
       await this._market.marketContractWrapper.getAllowanceAsync(
-        this._collateralTokenAddress,
+        this._market.mktTokenContract.address,
```
Comment (pelsasser) [inline src/order_watcher/RemainingFillableCalc.ts]: good catch!!

File: src/order_watcher/EventWatcher.ts (hunk 2 above)
Comment (perfectmak) [reply src/order_watcher/EventWatcher.ts]: True, I was really carried away with the implementation that I forgot about the docs. I add some docs to this PR before it gets merged.

File: src/contract_wrappers/ContractWrapper.ts (hunk 1 above)
Comment (perfectmak) [reply src/contract_wrappers/ContractWrapper.ts]: Yes, it would make it more uniform with other implementations. I'll update it. Thanks.

**Comment by perfectmak** [comment]
@pelsasser, I've pushed updates based on the review you made. Thanks.

**Pull request merged**